### ISSUE TITLE
diag(p3): off-PPO baseline (REINFORCE or A2C) — is the plateau PPO-specific or RL-general?

### ISSUE THREAD

**Issue by rjwalters** [rjwalters/bucket-brigade#273]
## Hypothesis being tested

After PPO + MAPPO both plateau, we don't know if this is PPO-specific (e.g., a clip-ratio or GAE pathology) or general to policy-gradient RL on this game. A different algorithm with the same env tells us which.

**Why this is newly urgent (2026-05-18)**: Several env-side and CTDE interventions have merged since this issue was filed — LOLA (#303), HCA (#301), COMA (#305), social-influence (#302), team-welfare shaping (#300), Δsafe progress shaping (#299). All of these still rely on PPO as the policy-gradient core. If PPO itself is the bottleneck, every CTDE/intrinsic-reward result above is currently confounded. A REINFORCE control discriminates "PPO clip-ratio pathology" from "policy-gradient RL general failure" and recalibrates how to read the new interventions.

## Protocol

1. Implement a minimal **REINFORCE** trainer (vanilla policy gradient, no advantage estimation, no clip, no GAE, no value baseline) in `bucket_brigade/training/reinforce_trainer.py`, mirroring `bucket_brigade/training/lola_trainer.py`'s API (`__init__`, `collect_rollout`, `update`).
2. Wire it into `experiments/p3_specialization/train.py` behind a CLI flag (`--algorithm reinforce`) on the same `if cfg.lola_dice: ... else:` trainer-switch (`train.py:840`).
3. Train for the **same total env steps** as IPPO baseline (PR #257) on `minimal_specialization`, 3 seeds (42, 43, 44).
4. Compare converged mean gap_closed against IPPO (~0.182 from #257) and against the new LOLA/HCA/COMA/social-influence baselines.

### Why REINFORCE over A2C

The original spec left both open. Recommend **REINFORCE**:

- **Fewer moving parts**: no critic, no advantage estimator, no value-loss weighting. If vanilla PG plateaus, the result is unambiguous — the failure isn't masked by a misspecified critic.
- **Smallest diff from PPO**: rollout collection machinery is reused as-is; only the `update()` step changes.
- **Matches the framing**: "vanilla policy gradient, no advantage estimation."
- **Smaller scope per builder economy**: ~150 LOC vs ~250 for A2C.

If REINFORCE is too high-variance to interpret, A2C is a natural follow-on (~30 LOC delta) — but starting there muddies the signal.

## Implementation sketch (~150 LOC)

Reuse, do not rewrite. Follow the `LolaTrainer` pattern (`bucket_brigade/training/lola_trainer.py` on origin/main, 723 LOC):

- **Policy network**: `bucket_brigade.training.networks.PolicyNetwork` unchanged. The existing `get_action_and_value()` returns a value head we simply discard.
- **Rollout collection**: mirror `JointPPOTrainer.collect_rollout()` (`bucket_brigade/training/joint_trainer.py:393`). Same `RolloutBuffer` shape minus values; log-probs and per-step rewards are sufficient. Import `RolloutBuffer` and `flatten_dict_obs` from `joint_trainer`, as `LolaTrainer` does.
- **Env wiring**: same `env_fn` pattern as `JointPPOTrainer.__init__` (`joint_trainer.py:193`).

New code only in `update()`:

```python
# Compute discounted returns per agent (no GAE, no bootstrap from V)
returns = torch.zeros_like(rewards)
G = torch.zeros(N)
for t in reversed(range(T)):
    G = rewards[t] + gamma * G * (1 - dones[t])
    returns[t] = G

# Optional: normalize returns to stabilize (orthogonal to advantage estimation)
returns = (returns - returns.mean()) / (returns.std() + 1e-8)

# Policy gradient: maximize log_prob * return
loss = -(log_probs * returns.detach()).mean()
```

**Explicitly omit**: clip ratio, GAE, value loss, multiple update epochs per rollout, KL penalty. One gradient step per rollout, on-policy only.

**Mutual exclusion** (mirror `lola_trainer.py:__init__` and `joint_trainer.py:236`): REINFORCE must reject `centralized_critic`, `redundancy_coef > 0`, `coma=True`, `advantage_estimator != "gae"` (treat GAE as the "off" sentinel), `influence_coef > 0`, `lola_dice=True`. Raise `ValueError` at construction if any of these are set. The whole point of this diagnostic is *clean vanilla PG*.

## Comparability constraints (critical for verdict)

To make the off-PPO comparison interpretable, hold these constant vs IPPO baseline (PR #257):

- Same `PolicyNetwork` hyperparameters (`hidden_size`, `num_layers`)
- Same scenario (`minimal_specialization`, 4 agents)
- Same rollout length (`num_steps`)
- Same `gamma`
- **Same total environment steps** (NOT iterations — REINFORCE does 1 update/rollout, PPO does many)
- 3 matched seeds: 42, 43, 44

Use `experiments/scenarios/minimal_specialization/` for config parity.

## Verdict logic

| Off-PPO outcome | Interpretation |
|---|---|
| **Same plateau as PPO** (~0.182 gap_closed) | Failure is RL-general, not PPO-specific. The plateau is a structural/representation/game-theoretic property of the env. Justifies doubling down on env-side fixes (the new CTDE/intrinsic interventions are the right axis); deprioritize PPO-knob tuning. |
| **REINFORCE > PPO** | PPO's clip ratio / GAE is hurting us specifically. Re-examine clip-ratio annealing, GAE λ extremes, ppo_epochs=1, num minibatches. Possibly explains why MAPPO didn't help — same PPO core. |
| **REINFORCE < PPO** (more chaotic) | PPO is mitigating something real (high variance, off-policy correction). The plateau is genuinely "best PPO can do here," but PPO is necessary. Search for orthogonal fixes (env, reward, init). |

## Affected files

- `bucket_brigade/training/reinforce_trainer.py` (NEW, ~150 LOC) — `REINFORCETrainer` class mirroring `LolaTrainer`'s API surface.
- `experiments/p3_specialization/train.py` — add `algorithm: str = "ppo"` to `CellConfig` (or reuse existing trainer-switch pattern at `train.py:840`); branch to `REINFORCETrainer` when `cfg.algorithm == "reinforce"`. Add mutual-exclusion guards mirroring the LOLA block at `train.py:716`.
- `experiments/p3_specialization/run_reinforce_baseline.py` (NEW, ~80 LOC) — sweep driver analogous to `experiments/p3_specialization/run_lola_sweep.py`. Anchor at IPPO baseline hyperparams; grid is just `seeds × {return_normalize: on, off}` = 6 cells.
- `experiments/p3_specialization/analyze_reinforce.py` (NEW, ~60 LOC) — verdict analyzer producing the 3-way table above; reads `metrics.json` per cell.
- `tests/training/test_reinforce_trainer.py` (NEW) — see Test Plan.
- **No edits expected**: `bucket_brigade/training/networks.py`, `joint_trainer.py`, `lola_trainer.py`, `experiments/scenarios/minimal_specialization/`.

## Test Plan

Bit-identity is not applicable (new file, no existing reference). Mirror `tests/training/test_lola_trainer.py` structure:

- [ ] **Smoke test** (`tests/training/test_reinforce_trainer.py::test_smoke`): 4 agents, 2 iterations on `minimal_specialization`, rollout length 32. Loss is finite, no NaN, every per-agent policy has non-zero gradient on at least one parameter.
- [ ] **Mutual exclusion** (`test_mutex`): trainer construction raises `ValueError` when paired with each of `centralized_critic=True`, `coma=True`, `redundancy_coef > 0`, `influence_coef > 0`, `advantage_estimator="hca"`, `lola_dice=True`.
- [ ] **Reproducibility** (`test_reproducibility`): same seed produces identical per-iteration losses within `1e-5` over 3 iterations.
- [ ] **IPD sanity** (`test_ipd_sanity`): on a standalone 2-agent iterated prisoner's dilemma driver (mirror `test_lola_trainer.py::test_ipd_sanity`), REINFORCE converges toward the always-defect equilibrium (mean reward ≈ -2.0) — this is the *positive control*: vanilla PG should fail to find cooperation in IPD, confirming the implementation has no accidental opponent-awareness or shaping leak.
- [ ] **Manual verification**: 50-iter smoke run on `minimal_specialization`, 3 seeds, on `$COMPUTE_HOST_PRIMARY` (NOT local — see Compute Requirements below). Compare mean episode reward and gap_closed vs IPPO baseline from PR #257.
- [ ] **Verdict recorded**: notebook entry under `experiments/p3_specialization/research_notebook/` mapping outcome to the 3-way verdict-logic table above.

## Compute requirements

**Do NOT run locally.** Per `CLAUDE.md` Compute Resource Guidelines, dispatch to `$COMPUTE_HOST_PRIMARY` (Mac Studio or equivalent from `.env`). REINFORCE is cheaper than PPO per step but the full sweep × 3 seeds × IPPO-matched total env steps is still ~hours of CPU. Use `tmux` for persistence.

## References

- Sister issues #270, #271, #272 (Phase 1 discriminators — all now closed; their verdicts inform but don't gate this work).
- PR #257 (IPPO baseline, ~0.182 gap_closed on `minimal_specialization`).
- PR #225 (MAPPO infra — orthogonal algorithm change that also plateaued).
- PR #303 (LOLA-DiCE trainer — the API pattern this issue follows).
- PRs #301, #302, #305 (HCA, social-influence, COMA — all PPO-core CTDE interventions whose interpretation depends on this control).
- `docs/ppo-help-brief.md`

## Acceptance criteria

- [ ] `bucket_brigade/training/reinforce_trainer.py` implementing `REINFORCETrainer` (≤ ~200 LOC).
- [ ] `experiments/p3_specialization/train.py` CLI integration via `--algorithm reinforce`.
- [ ] `tests/training/test_reinforce_trainer.py` covering smoke, mutex, reproducibility, IPD sanity (all 4 tests pass).
- [ ] Sweep driver + analyzer under `experiments/p3_specialization/`.
- [ ] 50-iter smoke comparison vs IPPO baseline on `$COMPUTE_HOST_PRIMARY`.
- [ ] Verdict per 3-way table, recorded in `research_notebook/`.

**Comment by rjwalters** [comment]
## Curator Enhancement

### Algorithm recommendation: **REINFORCE over A2C**

Both options were left open in the original spec. Recommending REINFORCE:

- **Fewer moving parts**: no critic, no advantage estimator, no value-loss weighting. If vanilla PG plateaus, the result is unambiguous — the failure isn't masked by a misspecified critic.
- **Smallest diff from PPO**: collect-rollout machinery is reused as-is; only the `update()` step changes.
- **Matches the framing in the issue body**: "vanilla policy gradient, no advantage estimation."

If REINFORCE is too high-variance to interpret, A2C is a natural follow-on (~30 LOC delta) — but starting there muddies the signal.

### Implementation sketch (~150 LOC)

Reuse, do not rewrite:
- **Policy network**: `bucket_brigade.training.networks.PolicyNetwork` unchanged. The existing `get_action_and_value()` returns a value head we simply discard; no need to add a value-headless variant.
- **Rollout collection**: mirror `JointPPOTrainer.collect_rollout()` (`bucket_brigade/training/joint_trainer.py:393`). Same `RolloutBuffer` shape minus values; log-probs and per-step rewards are sufficient.
- **Env wiring**: same `env_fn` pattern as `JointPPOTrainer.__init__` (`joint_trainer.py:193`).

New code only in `update()`:
```python
# Compute discounted returns per agent (no GAE, no bootstrap from V)
returns = torch.zeros_like(rewards)
G = 0.0
for t in reversed(range(T)):
    G = rewards[t] + gamma * G * (1 - dones[t])
    returns[t] = G

# Optional: normalize returns to stabilize (orthogonal to advantage estimation)
returns = (returns - returns.mean()) / (returns.std() + 1e-8)

# Policy gradient: maximize log_prob * return
loss = -(log_probs * returns.detach()).mean()
```

**Explicitly omit**: clip ratio, GAE, value loss, multiple update epochs per rollout, KL penalty. One gradient step per rollout, on-policy only.

### Comparability constraints (critical for verdict)

To make the off-PPO comparison interpretable, hold these constant vs IPPO baseline (PR #257):
- Same `PolicyNetwork` hyperparameters (hidden_size, num_layers)
- Same scenario (`minimal_specialization`)
- Same rollout length (`num_steps`)
- Same `gamma`
- Same total environment steps (not iterations — REINFORCE does 1 update/rollout, PPO does many)
- 3 matched seeds

Use `experiments/scenarios/minimal_specialization/` for config parity.

### Coordination: blocked on Phase 1

This is **Phase 2**, and the original issue explicitly says "execute only if Phase 1 (#270, #271, #272) is ambiguous." Applying `loom:blocked` until Phase 1 completes. The verdict logic from those three sister issues will determine whether REINFORCE is the right Phase 2 lever or whether the answer (basin trap vs. anti-attractor vs. representation failure) is already in hand.

## Dependencies

- [ ] #270: BC-init specialist then continue PPO — basin-trap vs anti-attractor
- [ ] #271: random-init MLP best-of-N vs trained PPO
- [ ] #272: specialist BC-fit alone — representation check

Execute only if Phase 1 verdicts are ambiguous.

## Affected Files

- `bucket_brigade/training/reinforce_trainer.py` (NEW, ~150 LOC) — `REINFORCETrainer` class mirroring `JointPPOTrainer` API surface (`collect_rollout`, `update`), minus critic/clip/GAE machinery.
- `bucket_brigade/training/networks.py` — read-only reuse of `PolicyNetwork` (no edits expected).
- `bucket_brigade/training/joint_trainer.py` — reference implementation only (no edits).
- `experiments/scripts/run_reinforce_baseline.py` (NEW) — driver script analogous to existing PPO experiment runners; matches `minimal_specialization` config from PR #257.
- `experiments/scenarios/minimal_specialization/` — reuse existing scenario config; no edits.
- `tests/training/test_reinforce_trainer.py` (NEW) — smoke test: 10-step rollout, single update, asserts loss decreases on a trivial bandit.

## Test Plan

- [ ] Manual verification: 50-iter smoke run on `minimal_specialization`, 3 seeds, on Mac Studio (`$COMPUTE_HOST_PRIMARY`). Compare mean episode reward vs IPPO baseline from PR #257.
- [ ] Automated tests: `tests/training/test_reinforce_trainer.py` — gradient flows, loss is finite, single update reduces loss on a constant-reward toy env.
- [ ] Edge cases: episodic boundary handling (returns reset on done), reward normalization on/off ablation (1 seed each).
- [ ] Verdict recorded: notebook entry mapping outcome to verdict-logic table in original issue body (same plateau / REINFORCE > PPO / REINFORCE < PPO).

**Comment by rjwalters** [comment]
Re-curating per user request. The original `loom:blocked` label had no explicit gate; refreshing against current trainers.

**Comment by rjwalters** [comment]
## Re-curation complete (2026-05-17)

Refreshed against current `main` after the env-side/CTDE interventions landed (#301 HCA, #302 social-influence, #303 LOLA-DiCE, #305 COMA, #300 team-welfare, #299 Δsafe).

**Material changes**:
- **Removed `loom:blocked`**: original label had no explicit Dependencies gate; Phase 1 sister issues (#270, #271, #272) are all closed. No longer blocked.
- **Sharpened relevance argument**: every new CTDE intervention above still uses PPO as the policy-gradient core. A REINFORCE control discriminates "PPO clip pathology" from "PG-RL-general failure" and disambiguates how to read those results.
- **Concrete API spec**: `REINFORCETrainer` follows the `LolaTrainer` pattern at `bucket_brigade/training/lola_trainer.py` (merged in #303). Cited file:line throughout body.
- **Mutual-exclusion guards** added explicitly: REINFORCE rejects centralized_critic, COMA, HCA, influence_coef, redundancy_coef, lola_dice — the diagnostic's value comes from being clean vanilla PG.
- **Test plan** mirrors `tests/training/test_lola_trainer.py` (smoke, mutex, reproducibility, IPD sanity-as-positive-control).
- **Affected files**: new reinforce_trainer.py, train.py `--algorithm` flag, sweep driver, analyzer, tests.
- **Compute note**: `$COMPUTE_HOST_PRIMARY`, not local.

Keeping `loom:curated`; not adding `loom:issue` (human approval gate). No duplicates found in open/closed issues.
